ansible/roles/test/files/ptftests/sad_path.py

```diff
@@ -0,0 +1,183 @@
+import datetime
+import re
+import subprocess
+import time
+
+from arista import Arista
+
+
+class PrebootTest(object):
+    def __init__(self, oper_type, vm_list, portchannel_ports, vm_dut_map, test_args, dut_ssh):
+        self.oper_type = oper_type
+        self.vm_list = vm_list
+        self.portchannel_ports = portchannel_ports
+        self.vm_dut_map = vm_dut_map
+        self.test_args = test_args
+        self.dut_ssh = dut_ssh
+        self.fails_vm = set()
+        self.fails_dut = set()
+        self.log = []
+        self.shandle = SadOper(self.oper_type, self.vm_list, self.portchannel_ports, self.vm_dut_map, self.test_args, self.dut_ssh)
+
+    def setup(self):
+        if 'bgp' in self.oper_type:
+            self.shandle.sad_setup(is_up=False)
+        return self.shandle.retreive_test_info(), self.shandle.retreive_logs()
+
+    def verify(self):
+        self.shandle.sad_bgp_verify()
+        return self.shandle.retreive_logs()
+
+    def revert(self):
+        self.shandle.sad_setup()
+        return self.shandle.retreive_logs()
+
+
+class SadPath(object):
+    def __init__(self, oper_type, vm_list, portchannel_ports, vm_dut_map, test_args):
+        self.oper_type = oper_type
+        self.vm_list = vm_list
+        self.portchannel_ports = portchannel_ports
+        self.vm_dut_map = vm_dut_map
+        self.test_args = test_args
+        self.neigh_vm = None
+        self.neigh_name = None
+        self.vm_handle = None
+        self.neigh_bgp = None
+        self.dut_bgp = None
+        self.log = []
+        self.fails = dict()
+        self.fails['dut'] = set()
+
+    def cmd(self, cmds):
+        process = subprocess.Popen(cmds,
+                                   shell=False,
+                                   stdout=subprocess.PIPE,
+                                   stderr=subprocess.PIPE)
+        stdout, stderr = process.communicate()
+        return_code = process.returncode
+
+        return stdout, stderr, return_code
+
+    def select_vm(self):
+        self.vm_list.sort()
+        # use the day of the month to select a VM from the list for the sad pass operation
+        vm_index = datetime.datetime.now().day % len(self.vm_list)
+        self.neigh_vm = self.vm_list.pop(vm_index)
+
+    def get_neigh_name(self):
+        for key in self.vm_dut_map.keys():
+            if self.vm_dut_map[key]['mgmt_addr'] == self.neigh_vm:
+                self.neigh_name = key
+                break
```

Comment:
You don't need a loop here
```
if key in self.vm_dut_map and self.vm_dut_map[key]['mgmt_addr'] == self.neigh_vm:
  self.neigh_name = key
```

Reply:
will change it

Reply:
done

Reply:
reverting back. I still need the loop. The keys in the vm_dut_map are the vm names. I need to extract the vm name based on the mgmt_addr of the down vm which is a key:value pair inside the vm name dict